### PR TITLE
Rewrite where with id

### PR DESCRIPTION
**Pull request by mad**
I try fix that issue https://github.com/opencypher/cypher-for-gremlin/issues/223

Query like that

```
MATCH (n)
WHERE id(n) in [1, 2]
RETURN n

MATCH (n)
WHERE id(n) = 1
RETURN n
```

may translate to:

```
g.V().has('~id', within(1, 2))...
g.V().has('~id', 1)...
```

### CODE REVIEW

File: translation/src/main/scala/org/opencypher/gremlin/translation/ir/rewrite/GroupStepFilters.scala [hunk 1]
```diff
@@ -93,6 +94,28 @@ object GroupStepFilters extends GremlinRewriter {
     traversals.flatMap {
       case SelectK(stepLabel) :: Values(propertyKey) :: Is(predicate) :: Nil =>
         (stepLabel, HasP(propertyKey, predicate)) :: Nil
+      case SelectK(stepLabel) :: ChooseP2(_, Id :: Nil) :: Is(_) :: Is(predicate) :: Nil =>
```
Comment (dwitry) [inline translation/src/main/scala/org/opencypher/gremlin/translation/ir/rewrite/GroupStepFilters.scala]: Good change, makes sense.

File: translation/src/main/scala/org/opencypher/gremlin/translation/ir/rewrite/GroupStepFilters.scala [hunk 2]
```diff
@@ -93,6 +94,28 @@ object GroupStepFilters extends GremlinRewriter {
...
+        (stepLabel, HasP(T.id.getAccessor, predicate)) :: Nil
+      case Project(_ @_*) :: rest =>
+        val values = rest.map {
```
Comment (dwitry) [inline translation/src/main/scala/org/opencypher/gremlin/translation/ir/rewrite/GroupStepFilters.scala]: This works, but we could make previous case to handle this case. Processing of within/without predicates [could be improved](https://github.com/dwitry/cypher-for-gremlin/commit/04b373bbae6d967abf3346db828a2a16a54d1c73) to provide cleaner Gremlin translation, that matches by `SelectK :: ChooseP2...` case you've added.

**Comment by dwitry** [comment]
Hello @mad, thank you for your contribution. I will improve translation to simplify your change as mentioned in previous comment. Then we will merge this.

**Comment by mad** [comment]
I found problem with UNWIND

```
UNWIND {data} as data 
MERGE (n {a: data.name})
SET n.name = data.name
RETURN id(n)
```

Translated to

```
g.inject('  cypher.start').choose(__.constant('  cypher.null'), __.constant('  cypher.null'), __.constant('  cypher.null')).unfold().as('data')
.choose(
__.V().as('n').where(__.project('  GENERATED4', '  GENERATED5').by(__.select('data')).by(__.constant('name')).select(values).map(cypherContainerIndex()).is(neq('  cypher.null')).as('  GENERATED3')
.select('n').values('a').where(eq('  GENERATED3'))),
 __.V().as('n').where(__.project('  GENERATED4', ' GENERATED5').by(__.select('data')).by(__.constant('name')).select(values).map(cypherContainerIndex()).is(neq('  cypher.null')).as('  GENERATED3')
.select('n').values('a').where(eq('  GENERATED3'))), __.identity().addV().as('n').property(single, 'a', __.project('  GENERATED1', '  GENERATED2').by(__.select('data')).by(__.constant('name')).select(values).map(cypherContainerIndex())))
.select('n').choose(neq('  cypher.null'), __.choose(__.project('  GENERATED6', '  GENERATED7').by(__.select('data')).by(__.constant('name')).select(values).map(cypherContainerIndex()).is(neq('  cypher.null')).unfold(), __.property(single, 'name', __.project('  GENERATED6', '  GENERATED7').by(__.select('data')).by(__.constant('name')).select(values).map(cypherContainerIndex())), __.sideEffect(__.properties('name').drop()))).select('n').project('id(n)').by(__.choose(neq('  cypher.null'), __.id()))
```

But extract `has(propperty, value` is not trivial. I think it's possible unfold unwind to multiple queries with constant and then GroupStepRewriter can rewrite it

File: translation/src/main/scala/org/opencypher/gremlin/translation/ir/rewrite/GroupStepFilters.scala (hunk 2 above)
Comment (mad) [reply translation/src/main/scala/org/opencypher/gremlin/translation/ir/rewrite/GroupStepFilters.scala]: Ok, I change it

**Comment by dwitry** [comment]
Hello @mad. Regarding `UNWIND`. You are right, theoretically it should be possible to extract translation of expression to variable [in WhereWalker](
https://github.com/opencypher/cypher-for-gremlin/blob/master/translation/src/main/scala/org/opencypher/gremlin/translation/walker/WhereWalker.scala#L69), to reduce amount of steps in `where` clause, therefore making it easier to match in `GroupStepRewriter`.

But its significant change, and I suggest to implement it in scope of other PR. If you agree we can merge this.

**Comment by mad** [comment]
Ok, I'm agree

**Comment by dwitry** [comment]
Closes #223

**Pull request merged**